File: sandbox/fortranpython/navaro/fortranarrays.f90
! from http://calcul.math.cnrs.fr/Documents/Ecoles/2010/f2py-cours.pdf
! Pierre Navaro


! tableaux:
!  - attention: passer le bon type sinon, f2py effectue une copie
!    et travaille sur une variable temporaire (sans rien dire!)
!  - si multidim: specifier "order='F'" pour le tableau numpy!

subroutine move( positions, vitesses, dt, n)
    integer, intent(in) :: n
    real(8), intent(in) :: dt
    real(8), dimension(n,3), intent(in) :: vitesses
    real(8), dimension(n,3) :: positions
    do i = 1, n
        positions(i,:) = positions(i,:) + dt*vitesses(i,:)
    end do
end subroutine move


! creation d'un tableau et retour a python.
! f2py se charge de la gestion de mémoire
subroutine create_array(A, n)
    integer, intent(in) :: n
    real(8), dimension(n,3), intent(out) :: A
    do i = 1, n
        do j = 1, n
            A(i,j) = i+j
        enddo
    enddo
end subroutine create_array

!
! produit matriciel
!

subroutine mult_array(A, B, C, m, n, k)

    real(8), dimension(m,k), intent(in) :: A
    real(8), dimension(k,n), intent(in) :: B
    integer :: m,n,k
    real(8), dimension(m,n), intent(out) :: C
    
    real(8) ::s
    integer :: i,j,l 

    do i=1, m 
        do j = 1, n 
            s = 0.0
            do l = 1, k
                s = s + A(i,l)*B(l,j)
            enddo
            C(i,j) = s
        enddo
    enddo
end subroutine mult_array

File: sandbox/fortranpython/navaro/fortranarrays.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import numpy as np 
import navaro as na

vitesses = np.array([[0, 1, 2], [0, 3, 2], [0, 1, 3]], dtype=np.float64)
positions = np.array([[0, 0, 0], [0, 0, 0], [0, 0, 0]], dtype=np.float64)
print('type(pos) =', positions.dtype)
na.move(positions, vitesses, 0.1)
print('p=', positions) #le tableau n'est pas mis a jour, stockage C

positions = np.array(positions, dtype=np.int8, order='F')
print('type(pos) =', positions.dtype)
na.move(positions, vitesses, 0.1)
print('p=', positions) #le tableau n'est pas mis a jour, mauvais type!

positions = np.array(positions, dtype=np.float64, order='F')
print('type(pos) =', positions.dtype)
na.move(positions, vitesses, 0.1)
print('p=', positions) #le tableau est modifie, stockage Fortran

##########

print('A=', na.create_array(5))

########

A = np.array([[0, 1, 2], [2, 3, 4]], dtype=np.float64)
B = np.array([[2, 1], [3, 2], [1, 9]], dtype=np.float64)
C = na.mult_array(A,B)
print("C=", C)
print("C=", np.dot(A,B))

